### ISSUE TITLE
Questions about GFLOPs calculation

### ISSUE THREAD

**Issue by WangQvQ** [ultralytics/ultralytics#7301]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

Looked at https://github.com/ultralytics/ultralytics/issues/6846, I found a very interesting question,

In your get_flops function, the value of stride is used to create a simulated input tensor im, which is then used to evaluate FLOPs  of the model. Specifically, stride determines the spatial dimension (i.e., width and height) of the analog input tensor.

![image](https://github.com/ultralytics/ultralytics/assets/58406737/92e2b6bc-98e8-4d2a-9c97-7deccba5e032)

Also, An error occurred when handling the rtdetr model: an error with selected index k out of range.

![image](https://github.com/ultralytics/ultralytics/assets/58406737/776a65b5-ddbf-474c-bd7b-862e389f9304)

In your get_flops function, using stride instead of directly using imgsz to set the input tensor size may indeed seem unusual. In general, calculations for FLOPs should be made based on the actual or expected input size of the model, which is often associated with imgsz.

So I tried to modify this method so that the RT-DETR model can also correctly print FLOPs. In addition, I have tested GFLOPs of YOLO detection model, segmentation model and key point model. The following is the specific table.

Model | original get_flops | My get_flops
-- | -- | --
YOLOv8n | 8.9 | 8.9
YOLOv8s | 28.8 | 28.8
YOLOv8m | 79.3 | 79.3
YOLOv8l | 165.7 | 165.7
YOLOv8x | 258.5 | 258.5
YOLOv8n-pose | 9.3 | 9.3
YOLOv8n-seg | 12.8 | 12.8
YOLOv8n-cls | 4.4 | 3.4
rt-detr-l | error | 108.3
rt-detr-x | error | 232.7

I think this part of the code seems to have some problems, if you think my idea is correct, you can consider my code.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by WangQvQ** [comment]
@glenn-jocher I don't know if my idea is correct, please have a look, thank you!

```
def get_flops(model, imgsz=640):
    """Return a YOLO model's FLOPs."""
    try:
        model = de_parallel(model)
        p = next(model.parameters())

        # Use actual image size for input tensor
        im = torch.empty((1, p.shape[1], imgsz, imgsz), device=p.device)  # input image in BCHW format

        # Calculate FLOPs
        flops = thop.profile(deepcopy(model), inputs=[im], verbose=False)[0] / 1E9 * 2 if thop else 0

        return flops  # Return FLOPs for the given image size
    except Exception as e:
        print(f"An error occurred: {e}")
        return 0
```

**Comment by glenn-jocher** [comment]
@WangQvQ yes we use 1 stride to evaluate flops rather than the full image size as i.e. at 640x640 and stride 32 then full image size with be 20 * 20 = 400 times slower to run through the GFLOPs computation. 

This is much, much slower, so our approach evaluates at the smallest possible size, then scales this evaluation to the full image size, which produces a FLOPs computation 400 times faster than your method normally. 

I would say you for the best combination of speed and robustness you could use a try: except statement to first attempt at stride 1, then if there is an error attempt again at the full image size. This should allow RTDETR FLOPs computations while keeping the current speed.

**Comment by glenn-jocher** [comment]
@WangQvQ hello! Thank you for diving into the FLOPs calculation and for your willingness to contribute with a PR. Your approach to directly use `imgsz` for setting the input tensor size is indeed more straightforward and aligns with the expected input size of the model. This can provide a more accurate representation of the computational complexity for a given input resolution.

Your findings and the table you provided are valuable, and it's great to see that the modified function can handle the RT-DETR model without errors. Your idea seems correct, and I appreciate your initiative to improve the codebase.

Please feel free to submit a PR with your proposed changes. We'll review it to ensure consistency and compatibility with the rest of the models. Your contribution is a testament to the collaborative spirit of the YOLO community. Thanks again! 🚀

**Comment by Ever-Bright-V** [comment]
> @glenn-jocher I don't know if my idea is correct, please have a look, thank you!
> 
> ```
> def get_flops(model, imgsz=640):
>     """Return a YOLO model's FLOPs."""
>     try:
>         model = de_parallel(model)
>         p = next(model.parameters())
> 
>         # Use actual image size for input tensor
>         im = torch.empty((1, p.shape[1], imgsz, imgsz), device=p.device)  # input image in BCHW format
> 
>         # Calculate FLOPs
>         flops = thop.profile(deepcopy(model), inputs=[im], verbose=False)[0] / 1E9 * 2 if thop else 0
> 
>         return flops  # Return FLOPs for the given image size
>     except Exception as e:
>         print(f"An error occurred: {e}")
>         return 0
> ```

@glenn-jocher @WangQvQ I am a beginner, why do floats need to be multiplied by 2。flops = thop.profile(deepcopy(model), inputs=[im], verbose=False)[0] / 1E9 * 2 if thop else 0

**Comment by Ever-Bright-V** [comment]
i found the comment in https://github.com/ultralytics/ultralytics/issues/7665#issue-2088163481. 
I calculated the GFLOPs of Restnet18 (input_size 224x224) using Flop to be 1.82e+09. This is similar to what was written in the article, so how should I handle this multiplication factor 2.
![68322b495d964ba9b35250d3f7c81ed3](https://github.com/ultralytics/ultralytics/assets/57551787/4fed0798-9ca4-4329-b075-aa41f3dc74a2)

> import numpy as np
> from thop import profile
> import torch
> import timm
> input = torch.randn(1, 3, 224, 224)
> model = timm.create_model('resnet18', pretrained=False)
> flops, params = profile(model, inputs=(input, ))
> # #
> print('{:.2e}'.format(flops))
> print('FLOPs = ' + str(flops/1000**3) + 'G')
> print('Params = ' + str(params/1000**2) + 'M')

**Comment by glenn-jocher** [comment]
@Ever-Bright-V hey there! 😊 The multiplication by 2 you're referring to is commonly used to account for both the multiply and accumulate operations in each floating-point operation (FLOP). Essentially, each FLOP consists of a multiplication followed by an addition. So, when calculating GFLOPs, we often multiply by 2 to include both these operations for a more accurate representation of the computational workload.

In your case with ResNet18, if you're following the convention used in many articles and benchmarks, you'd apply this factor to ensure consistency with those standards. However, it's always good to clarify the calculation method you're using when presenting your results, as practices can vary.

Here's a quick example to illustrate:
```python
# Assuming `flops` is the raw count from thop.profile()
gflops = flops / 1e9 * 2  # Convert to GFLOPs and account for multiply-accumulate operations
print(f'GFLOPs = {gflops}G')
```

Hope this clears things up! Let me know if you have any more questions.
